vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+
+A container that closest matches the semantics of R's named list is the
+`OrderedDict`, but that's relatively uncommon in Python code so we don't
+cover it further.
+
+#### Sets
+
+Sets are a container that can be used to efficiently track unique items
+or deduplicate lists. They are constructed using `{val1, val2}` (unlike
+a dictionary, without `:`). Think of them as dictionary where you only
+use the keys. Sets have many efficient methods for membership
+operations, like `intersection()`, `issubset()`, `union()` and so on.
+
+```{python}
+s = {1, 2, 3}
+type(s)
+s
+
+s.add(1)
+s
+```
+
+### Iteration with `for`
+
+The `for` statement in Python can be used to iterate over any kind of
+container.
+
+```{python}
+for x in [1, 2, 3]:
+  print(x)
+```
+
+R has a relatively limited set of objects that can be passed to `for`.
+Python by comparison, provides an iterator protocol interface, which
+means that authors can define custom objects, with custom behavior that
+is invoked by `for`. (We'll have an example for how to define a custom
+iterable when we get to classes). You may want to use a Python iterable
+from R using reticulate, so it's helpful to peel back the syntactic
+sugar a little to show what the `for` statement is doing in Python, and
+how you can step through it manually.
+
+There are two things that happen: first, an iterator is constructed from
+the supplied object. Then, the new iterator object is repeatedly called
+with `next()` until it is exhausted.
+
+```{python, error = TRUE}
+l = [1, 2, 3]
+it = iter(l) # create an iterator object
+it
+
+# call `next` on the iterator until it is exhausted:
+next(it)
+next(it)
+next(it)
+next(it)
+```
+
+In R, you can use reticulate to step through an iterator the same way.
+
+```{r}
+library(reticulate)
+l <- r_to_py(list(1, 2, 3))
+it <- as_iterator(l)
+
+iter_next(it)
+iter_next(it)
+iter_next(it)
+iter_next(it, completed = "StopIteration")
+```
+
+Iterating over dictionaries first requires understanding if you are
+iterating over the keys, values, or both. Dictionaries have methods that
+allow you to specify which.
+
+```{python}
+d = {"key1": 1, "key2": 2}
+for key in d:
+  print(key)
+
+for value in d.values():
+  print(value)
+
+for key, value in d.items():
+  print(key, ":", value)
+```
+
+#### Comprehensions
+
+Comprehensions are special syntax that allow you to construct a
+container like a list or a dict, while also executing a small operation
+or single expression on each element. You can think of it as special
+syntax for R's `lapply`.
+
+For example:
+
+```{python}
+x = [1, 2, 3]
+
+# a list comprehension built from x, where you add 100 to each element
+l = [element + 100 for element in x]
+l
+
+# a dict comprehension built from x, where the key is a string.
+# Python's str() is like R's as.character()
+d = {str(element) : element + 100
+     for element in x}
+d
+```
+
+### Defining Functions with `def`
+
+Python functions are defined with the `def` statement. The syntax for
```

Comment:
I think it's worth mentioning that Python functions require a `return` statement.

Reply:
Thanks! Good point.